File: ios/Classes/GamepadStreamHandler.swift
import Flutter

/// FlutterStreamHandler that forwards native gamepad events to Dart.
///
/// Events are queued if no sink is attached yet; once a listener subscribes
/// all queued events are flushed immediately.
public class GamepadStreamHandler: NSObject, FlutterStreamHandler {

    /// The active event sink provided by Flutter's EventChannel.
    private var eventSink: FlutterEventSink?

    /// Events received before the Dart side starts listening.
    private var pendingEvents: [[String: Any]] = []

    /// Thread-safety lock for sink / queue access.
    private let lock = NSLock()

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?,
                         eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        lock.lock()
        eventSink = events

        // Flush any events that arrived before the listener was attached.
        for event in pendingEvents {
            events(event)
        }
        pendingEvents.removeAll()
        lock.unlock()

        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        lock.lock()
        eventSink = nil
        lock.unlock()

        return nil
    }

    // MARK: - Internal API

    /// Sends a gamepad event dictionary to Dart.
    ///
    /// If the sink is not yet available the event is queued.
    func send(event: [String: Any]) {
        lock.lock()
        if let sink = eventSink {
            lock.unlock()
            // Dispatch on the main thread to satisfy Flutter platform channel
            // requirements.
            DispatchQueue.main.async {
                sink(event)
            }
        } else {
            pendingEvents.append(event)
            lock.unlock()
        }
    }
}

File: android/src/main/kotlin/dev/gamepad/GamepadStreamHandler.kt
package dev.gamepad

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * [EventChannel.StreamHandler] implementation that forwards native gamepad
 * events to the Dart side.
 *
 * Events received before a listener is attached are queued and flushed as soon
 * as [onListen] is called.
 */
class GamepadStreamHandler : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null

    /** Events received before the Dart side starts listening. */
    private val pendingEvents = mutableListOf<HashMap<String, Any>>()

    private val mainHandler = Handler(Looper.getMainLooper())

    private val lock = Any()

    // -- EventChannel.StreamHandler --------------------------------------------

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        synchronized(lock) {
            eventSink = events

            // Flush any events that arrived before the listener was attached.
            if (events != null) {
                for (event in pendingEvents) {
                    events.success(event)
                }
            }
            pendingEvents.clear()
        }
    }

    override fun onCancel(arguments: Any?) {
        synchronized(lock) {
            eventSink = null
        }
    }

    // -- Internal API ----------------------------------------------------------

    /**
     * Sends a gamepad event map to Dart.
     *
     * If the sink is not yet available the event is queued. Events are always
     * dispatched on the main (UI) thread to satisfy Flutter platform-channel
     * requirements.
     */
    fun send(event: HashMap<String, Any>) {
        synchronized(lock) {
            val sink = eventSink
            if (sink != null) {
                if (Looper.myLooper() == Looper.getMainLooper()) {
                    sink.success(event)
                } else {
                    mainHandler.post { sink.success(event) }
                }
            } else {
                pendingEvents.add(event)
            }
        }
    }
}
